core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -109,73 +109,91 @@ class TransactionCoordinator(brokerId: Int,
               state = Empty,
               topicPartitions = collection.mutable.Set.empty[TopicPartition],
-              lastUpdateTimestamp = time.milliseconds())
+              txnLastUpdateTimestamp = time.milliseconds())
 
-            val metadata = txnManager.addTransaction(transactionalId, newMetadata)
+            val epochAndMetadata = txnManager.addTransaction(transactionalId, newMetadata)
 
             // there might be a concurrent thread that has just updated the mapping
             // with the transactional id at the same time; in this case we will
             // treat it as the metadata has existed and update it accordingly
-            metadata synchronized {
-              if (!metadata.eq(newMetadata))
-                initPidWithExistingMetadata(transactionalId, transactionTimeoutMs, responseCallback, metadata)
+            epochAndMetadata synchronized {
+              if (epochAndMetadata.transactionMetadata.pendingTransitionInProgress)
+                responseCallback(initTransactionError(Errors.CONCURRENT_TRANSACTIONS))
+              else if (!epochAndMetadata.transactionMetadata.equals(newMetadata))
+                initPidWithExistingMetadata(transactionalId, transactionTimeoutMs, responseCallback, epochAndMetadata)
               else
-                appendMetadataToLog(transactionalId, metadata, responseCallback)
-
+                initPidWithNewMetadata(transactionalId, responseCallback, epochAndMetadata)
+            }
+          case Some(existingEpochAndMetadata) =>
+            existingEpochAndMetadata synchronized {
+              initPidWithExistingMetadata(transactionalId, transactionTimeoutMs, responseCallback, existingEpochAndMetadata)
             }
-          case Some(metadata) =>
-            initPidWithExistingMetadata(transactionalId, transactionTimeoutMs, responseCallback, metadata)
         }
       }
   }
 
-  private def appendMetadataToLog(transactionalId: String,
-                             metadata: TransactionMetadata,
-                             initPidCallback: InitPidCallback): Unit ={
+  private def initPidWithNewMetadata(transactionalId: String,
+                                     initPidCallback: InitPidCallback,
+                                     epochAndMetadata: CoordinatorEpochAndTxnMetadata): Unit = {
+    val metadata = epochAndMetadata.transactionMetadata
+
     def callback(errors: Errors): Unit = {
       if (errors == Errors.NONE)
         initPidCallback(initTransactionMetadata(metadata))
       else
         initPidCallback(initTransactionError(errors))
     }
-    txnManager.appendTransactionToLog(transactionalId, metadata, callback)
+
+    // set the pending state to Empty again to allow the metadata to be updated
+    // after it was successfully written to log
+    metadata.prepareTransitionTo(Empty)
+
+    txnManager.appendTransactionToLog(transactionalId, epochAndMetadata.coordinatorEpoch, epochAndMetadata.transactionMetadata, callback)
```

Comment:
I just refactored the code a bit around `initPid`, since it is a special case compared with others:

1. In other cases, we first create a new metadata as a place holder of all the pending updates to the original object and then only update it in the callback after append / send marker completes;

2. In `initPid` case, we would try first inserting a dummy into the cache but set its pending state, as `Empty -> Empty`; if there is an existing entry already we would either a) abort its existing txn first b) return retriable error code or c) update epoch / txn timeout and do the same as case 1) above.

The new code path looks better to me know. Let me know what do you think.